File: startup-backend/app/api/routes/scam_cases.py
import json
from uuid import uuid4

from fastapi import APIRouter, HTTPException
from sqlalchemy.orm import Session

from app.core.database import SessionLocal
from app.models.action_log import ActionLog
from app.models.scam_case import ScamCase
from app.schemas.scam_case import (
    CaseCloseUpdate,
    CaseNotesUpdate,
    CaseStatusUpdate,
    ScamCaseIntakeRequest,
    ScamCaseResponse,
)
from app.services.scam_case_logic import build_case_artifacts

router = APIRouter(prefix="/api/scam-cases", tags=["scam-cases"])


def write_action_log(db: Session, scam_case_id: int, action_type: str, details: str):
    log = ActionLog(
        scam_case_id=scam_case_id,
        action_type=action_type,
        details=details,
    )
    db.add(log)
    db.commit()
    db.refresh(log)
    return log


def serialize_action_logs(db: Session, scam_case_id: int) -> list[dict]:
    logs = (
        db.query(ActionLog)
        .filter(ActionLog.scam_case_id == scam_case_id)
        .order_by(ActionLog.created_at.desc(), ActionLog.id.desc())
        .all()
    )

    return [
        {
            "id": log.id,
            "created_at": log.created_at.isoformat() if log.created_at else None,
            "action_type": log.action_type,
            "details": log.details,
        }
        for log in logs
    ]


def serialize_case(db: Session, case: ScamCase) -> dict:
    return {
        "id": case.id,
        "case_id": case.case_id,
        "created_at": case.created_at.isoformat() if case.created_at else None,
        "updated_at": case.updated_at.isoformat() if case.updated_at else None,
        "status": case.status,
        "urgency": case.urgency,
        "urgency_score": case.urgency_score,
        "scam_type": case.scam_type,
        "title": case.title,
        "summary": case.summary,
        "customer_identifier": case.customer_identifier,
        "full_name": case.full_name,
        "age_band": case.age_band,
        "vulnerable_adult_flag": case.vulnerable_adult_flag,
        "source_unit": case.source_unit,
        "trusted_contact_exists": case.trusted_contact_exists,
        "trusted_contact_name": case.trusted_contact_name,
        "trusted_contact_phone": case.trusted_contact_phone,
        "intake_channel": case.intake_channel,
        "transaction_type": case.transaction_type,
        "amount_at_risk": case.amount_at_risk,
        "money_already_left": case.money_already_left,
        "customer_currently_on_call_with_scammer": case.customer_currently_on_call_with_scammer,
        "new_payee_or_destination": case.new_payee_or_destination,
        "customer_told_to_keep_secret": case.customer_told_to_keep_secret,
        "narrative": case.narrative,
        "urgency_reasons": json.loads(case.urgency_reasons_json),
        "risk_factors": json.loads(case.risk_factors_json),
        "playbook": json.loads(case.playbook_json),
        "notes": case.notes,
        "outcome_type": case.outcome_type,
        "closure_notes": case.closure_notes,
        "action_logs": serialize_action_logs(db, case.id),
    }


@router.get("/")
def list_scam_cases():
    db: Session = SessionLocal()

    try:
        cases = db.query(ScamCase).all()

        status_order = {"New": 0, "In Review": 1, "Escalated": 2, "Closed": 3}
        urgency_order = {"Critical": 0, "High": 1, "Medium": 2, "Low": 3}

        cases = sorted(
            cases,
            key=lambda case: (
                status_order.get(case.status, 99),
                urgency_order.get(case.urgency, 99),
                -(case.id or 0),
            ),
        )

        results = []
        for case in cases:
            results.append(
                {
                    "id": case.id,
                    "case_id": case.case_id,
                    "title": case.title,
                    "customer_identifier": case.customer_identifier,
                    "source_unit": case.source_unit,
                    "scam_type": case.scam_type,
                    "urgency": case.urgency,
                    "status": case.status,
                    "amount_at_risk": case.amount_at_risk,
                    "notes": case.notes,
                    "created_at": case.created_at.isoformat() if case.created_at else None,
                }
            )

        return results
    finally:
        db.close()


@router.post("/intake", response_model=ScamCaseResponse)
def create_scam_case(payload: ScamCaseIntakeRequest):
    db: Session = SessionLocal()

    try:
        artifacts = build_case_artifacts(payload)

        scam_case = ScamCase(
            case_id=f"CASE-{uuid4().hex[:8].upper()}",
            status="New",
            urgency=artifacts["urgency"],
            urgency_score=artifacts["urgency_score"],
            scam_type=artifacts["scam_type"],
            title=artifacts["title"],
            summary=artifacts["summary"],
            customer_identifier=payload.customer_identifier,
            full_name=payload.full_name,
            age_band=payload.age_band,
            vulnerable_adult_flag=payload.vulnerable_adult_flag,
            source_unit=payload.source_unit,
            trusted_contact_exists=payload.trusted_contact_exists,
            trusted_contact_name=payload.trusted_contact_name,
            trusted_contact_phone=payload.trusted_contact_phone,
            intake_channel=payload.intake_channel,
            transaction_type=payload.transaction_type,
            amount_at_risk=payload.amount_at_risk,
            money_already_left=payload.money_already_left,
            customer_currently_on_call_with_scammer=payload.customer_currently_on_call_with_scammer,
            new_payee_or_destination=payload.new_payee_or_destination,
            customer_told_to_keep_secret=payload.customer_told_to_keep_secret,
            narrative=payload.narrative,
            risk_factors_json=json.dumps(artifacts["risk_factors"]),
            urgency_reasons_json=json.dumps(artifacts["urgency_reasons"]),
            playbook_json=json.dumps(artifacts["playbook"]),
            notes="",
            outcome_type=None,
            closure_notes=None,
        )

        db.add(scam_case)
        db.commit()
        db.refresh(scam_case)

        write_action_log(
            db,
            scam_case.id,
            "case_created",
            f"Case created from {scam_case.source_unit} with {scam_case.urgency} urgency and scam type {scam_case.scam_type}.",
        )

        return serialize_case(db, scam_case)
    finally:
        db.close()


@router.get("/{case_id}", response_model=ScamCaseResponse)
def get_scam_case(case_id: int):
    db: Session = SessionLocal()

    try:
        case = db.query(ScamCase).filter(ScamCase.id == case_id).first()

        if not case:
            raise HTTPException(status_code=404, detail="Case not found")

        return serialize_case(db, case)
    finally:
        db.close()


@router.put("/{case_id}/status")
def update_scam_case_status(case_id: int, payload: CaseStatusUpdate):
    db: Session = SessionLocal()

    try:
        case = db.query(ScamCase).filter(ScamCase.id == case_id).first()

        if not case:
            raise HTTPException(status_code=404, detail="Case not found")

        old_status = case.status
        case.status = payload.status
        db.commit()
        db.refresh(case)

        write_action_log(
            db,
            case.id,
            "status_changed",
            f"Status changed from {old_status} to {case.status}.",
        )

        return {"id": case.id, "status": case.status}
    finally:
        db.close()


@router.put("/{case_id}/notes")
def update_scam_case_notes(case_id: int, payload: CaseNotesUpdate):
    db: Session = SessionLocal()

    try:
        case = db.query(ScamCase).filter(ScamCase.id == case_id).first()

        if not case:
            raise HTTPException(status_code=404, detail="Case not found")

        case.notes = payload.notes
        db.commit()
        db.refresh(case)

        preview = payload.notes.strip()[:120] if payload.notes.strip() else "Notes cleared."
        write_action_log(
            db,
            case.id,
            "notes_updated",
            preview,
        )

        return {"id": case.id, "notes": case.notes}
    finally:
        db.close()


@router.put("/{case_id}/close")
def close_scam_case(case_id: int, payload: CaseCloseUpdate):
    db: Session = SessionLocal()

    try:
        case = db.query(ScamCase).filter(ScamCase.id == case_id).first()

        if not case:
            raise HTTPException(status_code=404, detail="Case not found")

        case.status = "Closed"
        case.outcome_type = payload.outcome_type
        case.closure_notes = payload.closure_notes

        db.commit()
        db.refresh(case)

        write_action_log(
            db,
            case.id,
            "case_closed",
            f"Case closed with outcome {case.outcome_type}.",
        )

        return {
            "id": case.id,
            "status": case.status,
            "outcome_type": case.outcome_type,
            "closure_notes": case.closure_notes,
        }
    finally:
        db.close()

File: frontend/app/lib/scamCases.ts
export type BackendStatus = "New" | "In Review" | "Escalated" | "Closed";

export type ActionLog = {
  id: number;
  created_at?: string | null;
  action_type: string;
  details: string;
};

export type BackendScamCase = {
  id: number;
  case_id: string;
  created_at?: string | null;
  updated_at?: string | null;
  status: BackendStatus | string;
  urgency: "Low" | "Medium" | "High" | "Critical" | string;
  urgency_score: number;
  scam_type: string;
  title: string;
  summary: string;
  customer_identifier: string;
  full_name?: string | null;
  age_band: string;
  vulnerable_adult_flag: boolean;
  source_unit: string;
  assigned_owner?: string | null;
  assigned_team?: string | null;
  trusted_contact_exists: boolean;
  trusted_contact_name?: string | null;
  trusted_contact_phone?: string | null;
  intake_channel: string;
  transaction_type: string;
  amount_at_risk: number;
  money_already_left: boolean;
  customer_currently_on_call_with_scammer: boolean;
  new_payee_or_destination: boolean;
  customer_told_to_keep_secret: boolean;
  narrative: string;
  urgency_reasons: string[];
  risk_factors: Record<string, boolean>;
  playbook: {
    recommended_questions?: string[];
    recommended_actions?: string[];
    recommended_escalation_path?: string[];
    escalation_required?: boolean;
    hold_recommended?: boolean;
    trusted_contact_recommended?: boolean;
    law_enforcement_reporting_recommended?: boolean;
  };
  notes: string;
  outcome_type?: string | null;
  closure_notes?: string | null;
  action_logs: ActionLog[];
};

export type TimelineItem = {
  title: string;
  time: string;
  body: string;
};

export type QueueCase = {
  id: string;
  backendId?: number;
  caseNumber?: string;
  title: string;
  source: string;
  sourceGroup: string;
  member: string;
  risk: "Low" | "Medium" | "High" | "Critical";
  status: "New" | "Review" | "Escalated" | "Closed";
  owner: string;
  age: string;
  nextStep: string;
  summary: string;
  note: string;
  recommendedActions: string[];
  timeline: TimelineItem[];
};

export type ScamCaseIntakePayload = {
  customer_identifier: string;
  full_name?: string | null;
  age_band: string;
  vulnerable_adult_flag: boolean;
  source_unit: string;
  trusted_contact_exists: boolean;
  trusted_contact_name?: string | null;
  trusted_contact_phone?: string | null;
  intake_channel: string;
  transaction_type: string;
  amount_at_risk: number;
  money_already_left: boolean;
  customer_currently_on_call_with_scammer: boolean;
  new_payee_or_destination: boolean;
  customer_told_to_keep_secret: boolean;
  narrative: string;
  phone_based_imposter_story?: boolean;
  government_or_bank_brand_impersonation?: boolean;
  fear_or_urgency_language?: boolean;
  secrecy_pressure?: boolean;
  high_dollar_amount?: boolean;
  older_or_vulnerable_customer?: boolean;
  repeat_attempt?: boolean;
  remote_access_or_tech_support_story?: boolean;
  crypto_or_gift_card_request?: boolean;
  romance_or_emotional_dependency_pattern?: boolean;
};

const API_BASE = "/backend/api/scam-cases";

async function fetchJson<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...(init?.headers || {}),
    },
  });

  if (!response.ok) {
    let message = `Request failed with status ${response.status}`;
    try {
      const data = await response.json();
      message = data.detail || data.message || message;
    } catch {
      // Keep the generic message when the backend does not return JSON.
    }
    throw new Error(message);
  }

  return response.json() as Promise<T>;
}

export function listScamCases() {
  return fetchJson<BackendScamCase[]>(`${API_BASE}/`);
}

export function createScamCase(payload: ScamCaseIntakePayload) {
  return fetchJson<BackendScamCase>(`${API_BASE}/`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateScamCaseStatus(id: number, status: BackendStatus) {
  return fetchJson<BackendScamCase>(`${API_BASE}/${id}/status`, {
    method: "PUT",
    body: JSON.stringify({ status }),
  });
}

export function updateScamCaseAssignment(
  id: number,
  assigned_owner: string | null,
  assigned_team: string | null
) {
  return fetchJson<BackendScamCase>(`${API_BASE}/${id}/assignment`, {
    method: "PUT",
    body: JSON.stringify({ assigned_owner, assigned_team }),
  });
}

export function updateScamCaseNotes(id: number, notes: string) {
  return fetchJson<BackendScamCase>(`${API_BASE}/${id}/notes`, {
    method: "PUT",
    body: JSON.stringify({ notes }),
  });
}

export function deleteScamCase(id: number) {
  return fetchJson<{ id: number; deleted: boolean }>(`${API_BASE}/${id}`, {
    method: "DELETE",
  });
}

export function resetDemoData() {
  return fetchJson<{ deleted_cases: number; deleted_action_logs: number }>(
    `${API_BASE}/reset-demo-data`,
    { method: "POST" }
  );
}

function displayStatus(status: string): QueueCase["status"] {
  if (status === "In Review") return "Review";
  if (status === "Escalated" || status === "Closed" || status === "New") return status;
  return "New";
}

function sourceGroup(sourceUnit: string, intakeChannel?: string) {
  const source = `${sourceUnit} ${intakeChannel || ""}`.toLowerCase();
  if (source.includes("branch")) return "Branch network";
  if (source.includes("contact") || source.includes("call") || source.includes("phone")) return "Contact center";
  if (source.includes("digital") || source.includes("online")) return "Digital banking";
  return "Fraud ops";
}

function displayOwner(record: BackendScamCase) {
  const team = record.assigned_team || "";
  if (/fraud/i.test(team)) return "Fraud Ops";
  if (/member protection/i.test(team)) return "Member Protection";
  if (/branch/i.test(team)) return "Branch Ops";
  return record.assigned_owner || team || "Queue";
}

function relativeAge(value?: string | null) {
  if (!value) return "Today";

  const timestamp = new Date(value).getTime();
  if (Number.isNaN(timestamp)) return "Today";

  const minutes = Math.max(0, Math.round((Date.now() - timestamp) / 60000));
  if (minutes < 1) return "Just now";
  if (minutes < 60) return `${minutes} min`;

  const hours = Math.round(minutes / 60);
  if (hours < 24) return `${hours} hr`;

  return new Date(value).toLocaleDateString([], { month: "short", day: "numeric" });
}

function actionTitle(actionType: string) {
  const titles: Record<string, string> = {
    case_created: "Case created",
    status_changed: "Status changed",
    assignment_updated: "Assignment updated",
    notes_updated: "Operator note updated",
    case_closed: "Case closed",
  };
  return titles[actionType] || actionType.replace(/_/g, " ");
}

function timelineFromLogs(record: BackendScamCase): TimelineItem[] {
  if (!record.action_logs?.length) {
    return [
      {
        title: "Case available in operations",
        time: relativeAge(record.created_at),
        body: record.summary,
      },
    ];
  }

  return record.action_logs.slice(0, 6).map((log) => ({
    title: actionTitle(log.action_type),
    time: log.created_at
      ? new Date(log.created_at).toLocaleTimeString([], { hour: "numeric", minute: "2-digit" })
      : "Recent",
    body: log.details,
  }));
}

function nextStep(record: BackendScamCase) {
  if (record.status === "Closed") return "Closed with documentation";
  if (record.status === "Escalated") return "Supervisor and fraud review underway";
  if (record.status === "In Review") return "Awaiting documented follow-up and next action";

  const recommended = record.playbook?.recommended_escalation_path?.[0] || record.playbook?.recommended_actions?.[0];
  return recommended || "Document concern and move to first review.";
}

export function toQueueCase(record: BackendScamCase): QueueCase {
  const recommendedActions = record.playbook?.recommended_actions?.length
    ? record.playbook.recommended_actions
    : ["Review member intent and transaction urgency.", "Document observations and preserve timeline context."];

  return {
    id: String(record.id),
    backendId: record.id,
    caseNumber: record.case_id,
    title: record.title,
    source: record.source_unit,
    sourceGroup: sourceGroup(record.source_unit, record.intake_channel),
    member: record.full_name || record.customer_identifier,
    risk: ["Low", "Medium", "High", "Critical"].includes(record.urgency) ? (record.urgency as QueueCase["risk"]) : "Medium",
    status: displayStatus(record.status),
    owner: displayOwner(record),
    age: relativeAge(record.created_at),
    nextStep: nextStep(record),
    summary: record.summary || record.narrative,
    note: record.notes || "No operator note has been recorded yet.",
    recommendedActions,
    timeline: timelineFromLogs(record),
  };
}
